Watchdog refactoring

rajkan01 opened the pull request:
### Description

    Watchdog refactoring to multithreaded thread.
    -Added API to register muliple threads to watchdog drivers
    -Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
    -New APIs for watchdog
     wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
     wd_unregister(const osThreadId_t tid) to unregister to watchdog
     kick(const osThreadId_t tid) to refresh the watchdog
<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@mbed-os-hal

<!-- 
    Optional
    Request additional reviewers with @username
-->

Commit:
Remove window and sleep mode options for watchdog API

Commit:
Remove window and sleep mode options for watchdog API

Commit:
Remove window and sleep mode options for watchdog API

bulislaw committed:
Add RESET_REASON and WATCHDOG to doxygen options

bulislaw committed:
Disable Watchdog on Odin as it fails intermittently

Commit:
NCS36510: Added watchdog implementation.

ccli8 committed:
[Nuvoton] Support watchdog timer

ccli8 committed:
[Nuvoton] Support watchdog timer

Commit:
Added reset_reason feature for TMPM066 & TMPM46B

Commit:
Added reset_reason feature for TMPM066 & TMPM46B

ithinuel committed:
add the RFC for PinMap's extension

ithinuel committed:
add form_factors

tymoteuszblochmobica committed:
Multihoming initial release
Added Multihoming feature to LWIP (ability to use more than one network interfaces) for increasing networking reliability.
This involves:

    LWIP interface
    LWIP IP routing
    DNS storage
    Sockets (bind to interface name possibility)
    possibility to add non default network interface
    cellular middleware modifications if cellular connection is used

ithinuel committed:
fix types names.

ithinuel committed:
add precision about static arrays and fix functions' names

ithinuel committed:
update according to @bulislaw 's review

ithinuel committed:
fix according to @donatieng's review

Commit:
Add TF-M importer json

Commit:
Add platform TT_M4G9

Commit:
Add TF-M sources to .astyleignore

rajkan01 committed:
Watchdog refactoring to multithreaded thread.
-Added API to register muliple threads to watchdog drivers
-Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
-New APIs for watchdog
 wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
 wd_unregister(const osThreadId_t tid) to unregister to watchdog
 kick(const osThreadId_t tid) to refresh the watchdog

ghost commented:
@ARMmbed/mbed-os-hal

ciarmcom commented:
@rajkan01, thank you for your changes.
@ARMmbed/mbed-os-hal @ARMmbed/mbed-os-core @ARMmbed/mbed-os-maintainers please review.

cmonr commented:
@rajkan01 Please take a look at the `travis-ci/events` failure.

bulislaw commented on drivers/Watchdog.h:
```diff
@@ -34,10 +44,10 @@ namespace mbed {
  *
  * Watchdog watchdog = Watchdog();
- * watchdog.start(2000);
+ * watchdog.register(tid, 2000);
```
Why are we doing that? Do we have a use case for thread's support? I don't want force users to care about TID if they don't need to. It's a class why do we need `wd_` prefix? Please don't change the API names if it can be avoided. I know it's on feature branch and we could change the APIs, but best to avoid it if possible.

rajkan01 replied:
previous review comments on earlier version of watchdog driver requested multithreaded supported driver.
1.Need unique id to differentiate different thread while registering into watchdog driver, so plan to use tid as a key to differentiate multiple threads or if you have any suggestion we can discuss in the design meeting planned today
 2.we cant use start and stop API which is not match functionality because in the new implementation start and stop calls based on registered user/tid so plan to use API name as reg,unreg,kick without wd_.

deepikabhavnani replied:
@rajkan01 - I agree with @bulislaw that users should not be forced to pass thread ID. Using thread ID is great, but it can be done internally in driver using `get_id` of current thread https://github.com/ARMmbed/mbed-os/blob/master/rtos/ThisThread.h#L180.

It will also avoid thread1 - calling `kick` for thread2 and vice versa, as users can do this mistake. Also, if `get_id` is called from ISR context it will return NULL.

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+uint32_t Watchdog::_kick_bitmask = 0;
+Watchdog::watchdog_usr_info Watchdog::_usr_info[MAX_THREAD_WATCHDOG_SUPPORT] = {0};
+rtos::Mutex Watchdog::_ThreadSafeLockMutex;
+mbed_error_status_t Watchdog::wd_register(const osThreadId_t tid, const uint32_t timeout)
```
* No need for `wd_` prefix here
* Can you remove the `tid` parameter and use the current thread id instead? This will also provide us with a migration path for bare-metal support
* No need for `timeout` to be `const`

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+rtos::Mutex Watchdog::_ThreadSafeLockMutex;
+mbed_error_status_t Watchdog::wd_register(const osThreadId_t tid, const uint32_t timeout)
+{
+    int store_index = -1;
```
Could you define these variables closer to where they're needed?

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+    int store_index = -1;
+    watchdog_status_t sts;
+    int i;
+    _ThreadSafeLockMutex.lock();
```
`_mutex` is probably enough for name of that variable

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+    if(_is_initialized) {
+        if(timeout > conf_timeout) {
+            _re_initialize = 1;
+            hal_watchdog_stop();
```
Some watchdogs don't support re-init, also what happens if the MCU deadlocks here?

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+            hal_watchdog_stop();
+         }
+
+        for(i =0 ; i < MAX_THREAD_WATCHDOG_SUPPORT; i++) {
```
`i` should be instantiated here

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -32,22 +98,75 @@ watchdog_status_t Watchdog::start(const uint32_t timeout)
 
 
-void Watchdog::kick()
+void Watchdog::kick(const osThreadId_t tid)
```
Please remove `tid` as param, no point kicking from another thread

donatieng commented on drivers/Watchdog.h:
```diff
@@ -100,6 +115,47 @@ class Watchdog {
+    typedef struct
+    {
+       osThreadId_t tid;
+       uint32_t bit_idx;
```
Do you need this? Whatever position in the array can give the position in the bitmask?

donatieng commented on drivers/Watchdog.h:
```diff
@@ -100,6 +115,47 @@ class Watchdog {
+    mbed_error_status_t stop();
+
+private:
+    static rtos::Mutex _ThreadSafeLockMutex;//Thread safe mutex for concurrent access protection
```
How about using `PlatformMutex` which works in RTOS-less mode?

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -20,6 +20,72 @@
+            if(_bitmask & (1 << i))
+            {
+                if(_usr_info[i].tid == tid) {
+                    if(_re_initialize) {
```
Don't really understand this `_re_initialize` flag? Does this mean we fail if `timeout` < `conf_timeout`?

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -32,22 +98,75 @@ watchdog_status_t Watchdog::start(const uint32_t timeout)
+mbed_error_status_t Watchdog::wd_unregister(const osThreadId_t tid)
+{
+    int i;
+    uint32_t tbit = _bitmask;
```
Unsafe

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -32,22 +98,75 @@ watchdog_status_t Watchdog::start(const uint32_t timeout)
+            _usr_info[i].tid = 0;
+            _usr_info[i].bit_idx = 0;
+        }
+    }
```
You need to kick the hw watchdog if `_kick_bitmask == _bitmask`

donatieng commented on drivers/Watchdog.h:
```diff
@@ -100,6 +115,47 @@ class Watchdog {
+       uint32_t bit_idx;
+    }watchdog_usr_info ;
+
+    static watchdog_usr_info _usr_info[MAX_THREAD_WATCHDOG_SUPPORT];//Dats structure used to store the tid and their respective bit position of "_bitmask"
```
`usr_info` sounds vague

donatieng commented on drivers/Watchdog.h:
```diff
@@ -100,6 +115,47 @@ class Watchdog {
+    static watchdog_usr_info _usr_info[MAX_THREAD_WATCHDOG_SUPPORT];//Dats structure used to store the tid and their respective bit position of "_bitmask"
+
+    static uint32_t _bitmask; //Every bit is used to denote the registered threads/user
+    static uint32_t _kick_bitmask;//Every bit is used to denote the registered threads/user called kick(tid)
```
These should be bools

donatieng reviewed (changes requested):
Thanks @rajkan01 !
Additionally to my comments I feel some of this code could be made simpler. Atomics could be used to deal with `_bitmask` and `_kick_bitmask`.

caixue1102 committed:
Add WIFI support for RDA target UNO_91H

paul-szczepanek-arm committed:
add mtu events

Commit:
ONME-3949 Configuration options for STM_EMAC buffer counts

rajkan01 committed:
HwWatchdog role :
    - mbed_watchdog_mgr has interface name mbed_wdog_manager_start(),mbed_wdog_manager_stop(),mbed_wdog_manager_kick()
    - HwWatchdog is going to attach with LowPowerTIcker for periodic callback functionality
    - mbed_wdog_manager_start() will either get start either by BL/RTOS Aps,it reads the timeout value specified via macro and macro gets defined in target.json file.
    - mbed_wdog_manager_start() internally configure below HAL hw watchdog with timeout specified via target.json
    - mbed_wdog_manager_start() internally divide the timeout(specified in target.json) by the 2 and attach LowPowerTicker with periodic callback of hw_kick()
    - mbed_wdog_manager_start() internally create one instance of sw watchdog class,to access the static list data structure of sw watchdog class
    - mbed_wdog_manager_kick() function periodically get called and refresh the hw watchdog to avoid watchdog reset
    - converted C++ code into C based APIs
    - added boolean to control watchdog start and stop
    - Added detach from ticker on stop API

paul-szczepanek-arm committed:
route ble events to gap

paul-szczepanek-arm committed:
fix typos

rajkan01 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,89 @@
+      _Mutex.unlock();
+      return MBED_ERROR_OPERATION_PROHIBITED;
+    }
+    _Mutex.lock();
```
I have changed to unlock. Thanks for review comment

rajkan01 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,89 @@
+     }
+    else{
+        float timeout = (HW_WATCHDOG_TIMEOUT /2);
+        _ticker.attach(callback(this,&HwWatchdog::kick),(timeout / 1000));
```
LowPowerTicker works only millisecond accuracy ,If I change to attach_us ,its not working.Thanks for review comment

rajkan01 replied:
Yes I tried but its not worked

rajkan01 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,89 @@
+        _Mutex.unlock();
+        return MBED_ERROR_INITIALIZATION_FAILED;
+     }
+    else{
```
I have removed the else.Thanks for review comment

deepikabhavnani commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,84 @@
+}
+
+
+void HwWatchdog::kick()
```
Do we need this function to be user API? It is attached to ticker and used internally. 

Can we do `_ticker.attach(callback(this,&hal_watchdog_kick),(timeout / 1000));` and delete this API?

rajkan01 replied:
in the new implementation converted all APIs into C++ to C as suggested by Russ ,kick function is static and get accessed local to the file.Kick function gets extended future to go through the static list due to that we want to register kick into ticker.attach

deepikabhavnani commented on drivers/HwWatchdog.h:
```diff
@@ -0,0 +1,118 @@
+#ifdef DEVICE_WATCHDOG
+
+#include "watchdog_api.h"
+
```
Why do we need blank lines between includes. Nothing against it, but seemed odd

rajkan01 replied:
Yes removed unnecessary spaces

deepikabhavnani commented on drivers/HwWatchdog.h:
```diff
@@ -0,0 +1,118 @@
+ * @code
+ *
+ * HwWatchdog watchdog;
+ * watchdog.start();
```
example shows just `start` API for watchdog, should we be describing other uses cases as well in example?

rajkan01 replied:
once we added the sw watchdog and interface between HW watchdog to SW watchdog ,later we will update/add more description

deepikabhavnani commented on drivers/HwWatchdog.h:
```diff
@@ -0,0 +1,118 @@
+     *
+     * @return Reload value for the watchdog timer in milliseconds.
+     */
+    uint32_t reload_value() const;
```
When will user call this API? is it get API or set API? reload_value() - looks like user is reloading the value

Should it be - get_reload_value or get_refresh_value?

rajkan01 replied:
This API required to user to read currently configured timeout value in  hw watchdog, this naming convention used earlier, I am just used it

deepikabhavnani replied:
@c1728p9 @donatieng - reload_value - should it be get_reload_value?

deepikabhavnani commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,84 @@
+
+uint32_t HwWatchdog::max_timeout()
+{
+    const watchdog_features_t features = hal_watchdog_get_platform_features();
```
Are we returning a structure in return value?

deepikabhavnani replied:
Returning structure from hal API? Not sure how this was missed
https://github.com/ARMmbed/mbed-os/blob/feature-watchdog/hal/watchdog_api.h#L137 

CC @donatieng @c1728p9

rajkan01 replied:
we are returning only the max_timeout

Commit:
Doxygen correction

Adding to proper group so that the API doxygen appears into the class
hierarchy group rather than data strutures.

Commit:
mbed_retarget changes

- Avoid compiling user_setup_stackheap for TF-M secure targets
- Align sbrk for TF-M

paul-szczepanek-arm committed:
handle mtu and data len events for softdevice

paul-szczepanek-arm committed:
fix style

paul-szczepanek-arm committed:
change event names

Commit:
Add "SPDX-License-Identifier: Apache-2.0"

Commit:
Update code

Commit:
Cellular: Non-IP doxygen updates

paul-szczepanek-arm committed:
fix possible truncation

c1728p9 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,100 @@
+
+    watchdog_status_t sts;
+    mbed_error_status_t msts = MBED_SUCCESS;
+    MBED_STATIC_ASSERT(HW_WATCHDOG_TIMEOUT >= 0,"Timeout must be greater than zero");
```
This is a compile time assert, so you may want to move it outside of `start`. That way it is clear it doesn't run as part of `start`.

c1728p9 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,100 @@
+    return msts;
+}
+
+mbed_error_status_t stop()
```
Could this be simplified to a bool, with `true` meaning the watchdog was stopped and `false` meaning watchdog could not be stopped?

c1728p9 commented on drivers/HwWatchdog.h:
```diff
@@ -0,0 +1,94 @@
+/* mbed Microcontroller Library
```
These files would probably make more sense in the platform directory with a name like `mbed_watchdog_mgmt.h`. Camel case is used for classes and the file that contains them while for non-class code the convention is to use underscores.

c1728p9 commented on drivers/HwWatchdog.h:
```diff
@@ -0,0 +1,94 @@
+ *                 MBED_ERROR_INITIALIZATION_FAILED if one of the enabled input
+ *                 parameters is not supported by the current platform.
+ */
+mbed_error_status_t start();
```
Since these functions are in the global namespace you'll want to prefix them so they don't conflict with user code which may be using common names such as `start` and `stop`. If you rename this file as mentioned above you could name this something like `mbed_wdog_management_start()`.

c1728p9 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,100 @@
+    mbed_error_status_t msts = MBED_SUCCESS;
+    MBED_STATIC_ASSERT(HW_WATCHDOG_TIMEOUT >= 0,"Timeout must be greater than zero");
+    MBED_ASSERT(HW_WATCHDOG_TIMEOUT < max_timeout());
+    if(is_watchdog_started)
```
You may want to do this check inside the critical section to prevent possible race conditions with two threads calling start for the first time at the same time.

c1728p9 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,100 @@
+
+static LowPowerTicker ticker;//tiker for periodic delay call back for kick()
+static bool is_watchdog_started = false; //boolean to control watchdog start and stop
+#define MS_TO_US(x) (x * 1000) //macro to convert millisecond to microsecond 
```
you may want to wrap `x` in parentheses to ensure `x` is evaluated fully before doing the multiplication.

c1728p9 commented on drivers/HwWatchdog.cpp:
```diff
@@ -0,0 +1,100 @@
+
+#include "HwWatchdog.h"
+
+static LowPowerTicker ticker;//tiker for periodic delay call back for kick()
```
A singleton pointer or getter function should be used here to ensure this does not take up RAM when it is not used. Something like
```C++
SingletonPtr<LowPowerTicker> ticker;
```

or
```C++
static LowPowerTicker *get_ticker()
{
    static LowPowerTicker ticker;
    return &ticker;
}
```

paul-szczepanek-arm committed:
add API to trigger MTU negotiation

c1728p9 commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,94 @@
+ *                 MBED_ERROR_INITIALIZATION_FAILED if one of the enabled input
+ *                 parameters is not supported by the current platform.
+ */
+mbed_error_status_t mbed_wdog_manager_start();
```
Can this be a bool return value like `mbed_wdog_manager_stop`?

c1728p9 commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,107 @@
+    MBED_ASSERT(HW_WATCHDOG_TIMEOUT < mbed_wdog_manager_get_max_timeout());
+    core_util_critical_section_enter();
+    if(is_watchdog_started)
+        return MBED_ERROR_OPERATION_PROHIBITED;
```
You'll need to call `core_util_critical_section_exit()` before returning.

c1728p9 commented on targets/targets.json:
```diff
@@ -34,6 +34,11 @@
                 "value": "0x0fffffff",
                 "macro_name": "MPU_ROM_END"
+            },
+			"hw-watchdog_timeout": {
```
Spaces should be used instead of tabs.

c1728p9 reviewed (changes requested):
Unintentionally marked approve

Commit:
Align existing partitions to work with TF-M

- ITS
- Crypto
- Platform

rajkan01 committed:
SW Watchdog role:
        -SW watchdog has interface name start(),stop(),kick()	Sw watchdog internally has static list and shared across multiple instance of SW watchdog
	- Sw watchdog initialize timeout value,unique string via constructor whenever threads created sw watchdog object
	-Threads make sure pass proper timeout value,Unique string while creating the instance.
	-start() called by components(BLE,WIFI etc.,),it adds the entry into static list with few details current count ,etc.,
	-kick() called by registered components(BLE,WIFI etc.) to reset current count to zero.
        -is_alive - interface API to mbed_watchdog_manager
        -implementation optimization

0xc0170 commented on drivers/Watchdog.cpp:
```diff
@@ -20,46 +20,118 @@
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
+    Watchdog *node;
+    node = first;
+    if(Watchdog::first == NULL)
```
Please run astyle for your changed files with our config

0xc0170 commented on drivers/Watchdog.h:
```diff
@@ -82,24 +75,51 @@ class Watchdog {
+      *                 successfully. false if tid is not registered
+      *                 and try to do unregister
+      */
+    bool  rm_from_list();
```
`remove_from_list()` rather

0xc0170 commented on drivers/Watchdog.h:
```diff
@@ -15,26 +15,28 @@
-#ifndef MBED_WATCHDOG_H
-#define MBED_WATCHDOG_H
+#ifndef MBED_SW_WATCHDOG_H
+#define MBED_SW_WATCHDOG_H
```
what does SW stand for here? I haven't noticed this in drivers header files

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,126 @@
+    } else {
+        while(node->next != NULL) {
+            node = node->next;
+            if(node->_tid == _tid) {
```
Could `this` be checked rather than `_tid`? Multiple watchdogs may be used on the same thread if multiple services are running on the same event queue.

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,126 @@
-    if (timeout > max_timeout()) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
-    }
+mbed_error_status_t Watchdog::start()
```
Could this be void instead of returning a status?

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,126 @@
 void Watchdog::kick()
 {
-    hal_watchdog_kick();
+    Watchdog *cur_ptr = first;
```
why is `first` read outside the critical section?

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,126 @@
 
-
-watchdog_status_t Watchdog::stop()
+mbed_error_status_t Watchdog::stop()
```
Could this be void instead of returning a status?

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,126 @@
+    }
 
-    return hal_watchdog_init(&config);
+    _tid = rtos::ThisThread::get_id();
```
Why does the watchdog keep track of thread IDs?

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,103 @@
+    core_util_critical_section_enter();
+    msts = remove_from_list();
+    if(msts) {
+        _is_initialized = false;
```
Should `_is_initialized` be set to false unconditionally? `remove_from_list` returning false means that this object wasn't in the list in the first place right?

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,103 @@
+Watchdog *Watchdog::first;
 
-watchdog_status_t Watchdog::start(const uint32_t timeout)
+bool Watchdog::add_to_list()
```
If you use a flag to keep track of if an instance has been added to the list, such as `_is_initialized`, then you don't need to loop through the list. You can add this element immediately to the front of the list.

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,103 @@
-    return features.max_timeout;
+Watchdog::~Watchdog()
+{
+    stop();
```
This should check `_is_initialized` so you don't get an assert if an object is destroyed without a call to `start()`.

c1728p9 commented on drivers/Watchdog.h:
```diff
@@ -15,26 +15,29 @@
 
-#ifndef MBED_WATCHDOG_H
-#define MBED_WATCHDOG_H
+#ifndef MBED_SW_WATCHDOG_H
```
You could probably remove `SW` from this name since this is just called watchdog.

c1728p9 commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,109 @@
+static void mbed_wdog_manager_kick()
+{
+    core_util_critical_section_enter();
+    hal_watchdog_kick();
```
This will need to be looping through the software watchdog list and checking for timeouts there as well.

c1728p9 commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,74 @@ namespace mbed {
+    bool  remove_from_list();
+private:
+    uint32_t _max_timeout; //_max_timeout initialized via constructor while creating instance of this class
+    std::string str;//To store the details of user
```
Using the string class here will use dynamic memory and pull in extra code. I recommend using a static C string instead as the Thread class does.

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,83 @@
-    config.timeout_ms = timeout;
+void Watchdog::add_to_list()
+{
+    this->next = first;
```
May want to skip adding it to the list if `_is_initialized` is set, as this means you are already in the list.

rajkan01 committed:
HW Watchdog and SW watchdog interface
  - Added Hw watchdog periodic kick calls software watchdog is_alive check

Commit:
Call tfm_ns_lock_init() on TFM NSPE boot

Commit:
Modify license

rajkan01 committed:
Hw Watchdog UnitTest cases
    -Added the unit test case for testing Hw watchdog
    -Added the supported stubs files

Commit:
ESP8266: handles reset ready as OOB

Makes possible to recover from spurious resets addiotionally to planned
ones.

Commit:
ESP8266: fixes network status refresh procedure

Commit:
ESP8266: does OOB handling based on SIGIO and not on a certain interval

Commit:
Initialize platform in trng test

Add calls to `mbedtls_platform_setup()` and
`mbedtls_platform_terminate()` to the trng greentea test, to
initialize the hardware acceleration engines, in some platforms.

rajkan01 committed:
SW watchdog unittest cases
  -Added the mock class function to mock mbed_assert_internal
  -Added the unit test case to test start,kick,stop
  -Modified the interface api name from is_alive to process
  -added the unit test cases for process

c1728p9 commented on UNITTESTS/drivers/Watchdog/test_watchdog.cpp:
```diff
@@ -0,0 +1,114 @@
+    mbed::Watchdog *watchdog = new mbed::Watchdog(500,"watchdog_unittest");
+    EXPECT_TRUE(watchdog);
+    EXPECT_CALL(mwatchdog, process()).Times(AtLeast(1));
+    watchdog->start();
```
What is this testing? The test is called `wdog_start_fail` but I don't see what the test failure condition is.

c1728p9 commented on UNITTESTS/drivers/Watchdog/test_watchdog.cpp:
```diff
@@ -0,0 +1,114 @@
+    watchdog->stop();
+    delete watchdog;
+}
+
```
It may be good to add a test to confirm that the watchdog fire at the correct time if kick is missed.

c1728p9 commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,83 @@
+        if(cur_ptr->_current_count > cur_ptr->_max_timeout) {
+            system_reset();
+        } else {
+            cur_ptr->_current_count++;
```
`_max_timeout` is in millisecond right? If so you'll want to increment `_current_count` more than on. You may also want to rename `is_alive` to indicate that it is more than just a status check; calling is_alive multiple times has side effects unlike a simple status check. Maybe something like `Watchdog::process`.  What do you think?

c1728p9 commented on drivers/Watchdog.h:
```diff
@@ -20,21 +20,24 @@
  *
- * Watchdog watchdog = Watchdog();
- * watchdog.start(2000);
+ * Watchdog watchdog = Watchdog(300,"Software Watchdog");
```
You may want to do direct initialization rather than assign. Something like `Watchdog watchdog(300,"Software Watchdog");`.

Commit:
Adding TF-M integration guide

ytsuboi committed:
[Wio BG96] Fix onboard modem initialization

Fix for issue #9450 by adding turning off the modem power at startup, so that modem is sure to be in the initial state.

ytsuboi committed:
[Wio 3G] Fix onboard modem initialization

Fix for issue #9450 by adding turning off the modem power at startup, so that modem is sure to be in the initial state.
Fix for wrong filename and classname.

// WIP

fkjagodzinski commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,93 @@
+ *
+ * @return Reload value for the watchdog timer in milliseconds.
+ */
+uint32_t mbed_wdog_manager_get_reload_value();
```
```suggestion
uint32_t mbed_wdog_manager_get_timeout();
```
To stay consistent with  `mbed_wdog_manager_get_max_timeout()`.

fkjagodzinski commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,111 @@
+
+MBED_STATIC_ASSERT((HW_WATCHDOG_TIMEOUT >= 0),"Timeout must be greater than zero");
+
+Watchdog watchdog(0,"Platform Watchdog");
```
Please use `astyle` formatting.

donatieng replied:
+1

fkjagodzinski replied:
It took me a while to figure out this instance will not cause a reset (with its `0` timeout and not being kicked) because it's never started and appended to the list. Could you add a comment explaining this is not intended to start and only provide the `process()` method?

paul-szczepanek-arm committed:
fix type name

fkjagodzinski commented on drivers/Watchdog.h:
```diff
@@ -20,21 +20,24 @@
+#include "rtos/ThisThread.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_power_mgmt.h"
+#include <string>
```
I think we don't need `<string>` here.

fkjagodzinski commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,67 @@ namespace mbed {
 public:
     Watchdog() {}
 
+    Watchdog(uint32_t timeout = 0,const char *str = NULL):_max_timeout(timeout),name(str),_current_count(0),_is_initialized(false),next(NULL) {}
```
The default `timeout` value should not be `0`.

donatieng replied:
Can the implementation be in the `.cpp` file?

fkjagodzinski commented on UNITTESTS/drivers/Watchdog/test_watchdog.cpp:
```diff
@@ -0,0 +1,149 @@
+
+TEST_F(TestWatchdog, wdog_constructor)
+{
+    mbed::Watchdog *watchdog = new mbed::Watchdog(500,"watchdog_unittest");
```
You should also test constructor with the default params `Watchdog()`.

donatieng replied:
Can this be instantiated on the stack instead? Why on the heap?

fkjagodzinski reviewed (commented):
Since the changes from the first commit are updated by the following commits it might be a good idea to squash them making the history more readable.

donatieng commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,67 @@ namespace mbed {
+      *
+      */
+    void remove_from_list();
+private:
```
Could you make sure all members are prefixed with `_`?

donatieng commented on platform/mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,111 @@
+static bool is_watchdog_started = false; //boolean to control watchdog start and stop
+#define MS_TO_US(x) ((x) * 1000) //macro to convert millisecond to microsecond
+
+MBED_STATIC_ASSERT((HW_WATCHDOG_TIMEOUT >= 0),"Timeout must be greater than zero");
```
Actually, this should be strictly greater than 0

donatieng commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,67 @@ namespace mbed {
+class Watchdog
+{
 public:
     Watchdog() {}
```
There's a bunch of uninitialised members if we use this one

donatieng replied:
Also the compiler won't know when to call this as you have default argument values in the other constructor

donatieng commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,67 @@ namespace mbed {
+     * Otherwise, the system is reset.
      */
-    static uint32_t max_timeout();
+    void process();
```
It's not really clear what this method does from the doc

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,83 @@
+        if(cur_ptr->_current_count > cur_ptr->_max_timeout) {
+            system_reset();
+        } else {
+            cur_ptr->_current_count += ((HW_WATCHDOG_TIMEOUT/2));
```
`HW_WATCHDOG_TIMEOUT / 2` does not mean anything in this context, it should be passed as parameter instead

donatieng replied:
Also should be changed to `(HW_WATCHDOG_TIMEOUT + 1) / 2` for the extreme case of `HW_WATCHDOG_TIMEOUT == 1`

donatieng reviewed (changes requested):
Thanks a lot @rajkan01 that's great stuff. Re-reading it I found a couple of minor issues.

fkjagodzinski commented on drivers/Watchdog.h:
```diff
@@ -44,62 +47,67 @@ namespace mbed {
+    void remove_from_list();
+private:
+    uint32_t _max_timeout; //_max_timeout initialized via constructor while creating instance of this class
+    const char *name;//To store the details of user
```
I can't see the `name` used anywhere. Do we need that?

rajkan01 commented:
The next phase we will add to collect the details about which one did not called kick
So Yes we will use this “name” for future to store the user details before do system reset

fkjagodzinski commented on drivers/Watchdog.cpp:
```diff
@@ -18,48 +18,83 @@
+Watchdog::~Watchdog()
+{
+    if(_is_initialized)
+        stop();
```
Missing braces in the `if` statement.

fkjagodzinski commented on UNITTESTS/platform/mbed_watchdog_mgr/test_mbed_watchdog_mgr.cpp:
```diff
@@ -0,0 +1,70 @@
+
+TEST_F(TestMbedWatchdogMgr, test_mbed_watchdog_mgr_start_fail)
+{
+    EXPECT_FALSE(mbed_wdog_manager_start());
```
We could be more verbose here and enclose the test logic in each test case rather than rely on the execution order of the test cases.

rajkan01 commented:
I have checked with Russ regarding extending the testcase to more robust ..but we decided this part in enough to proceed now …..

Commit:
Adding MBEDTLS_ENTROPY_NV_SEED macro check in device key to allow key derivation when there is no TRNG but there is DRBG

fkjagodzinski commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,92 @@
+ * watchdog timers if supported by the current platform.
+ *
+ * @return Returns true if the watchdog timer was succesfully
+ *         stopped, or if the timer was never started. Returns
```
Current implementation returns `false` if the timer was never started.

fkjagodzinski commented on platform/mbed_watchdog_mgr.h:
```diff
@@ -0,0 +1,92 @@
+extern "C" {
+#endif
+
+/** \addtogroup drivers */
```
This is outdated; the code is in the `platform/` now.

paul-szczepanek-arm committed:
route mtu changed events to server and client instead of gap

cmonr commented:
@rajkan01 I'm assuming that once this PR passes CI, it will be used to replace the current `feature-watchdog` branch, correct?

rajkan01 commented on UNITTESTS/drivers/Watchdog/test_watchdog.cpp:
```diff
@@ -0,0 +1,142 @@
+
+TEST_F(TestWatchdog, wdog_constructor_with_default_value)
+{
+    mbed::Watchdog watchdog;
```
This is to make sure able to create watchdog instance without parameter.

rajkan01 commented on drivers/Watchdog.cpp:
```diff
@@ -0,0 +1,110 @@
+
+Watchdog *Watchdog::_first = NULL;
+
+Watchdog::Watchdog(uint32_t timeout, const char *const str): _name(str)
```
Make sure str keep as constant while referred by _name pointer. and keep the other member initialization in definition to make more readability of the code.

Commit:
Including CTest explicitly

enable_test() seems to include CTest framework in the CMakeLists.txt but
that would result in inconsistencies and CTest framework will not be
able to find DartConfiguration file or MemCheck tool.

Including CTest framework explicitely in the CMakeLists.txt seems to
solve the issue.

Commit:
Proper size checking for link ADR commands

In a specific branch path 'adr_settings' in link_adr_request() API, the
structure adr_settings of type link_adr_params_t will be rendered
uninitialized. To prevent this we initialize the construct as zero.
In addition to that, to handle the case properly we should check for the
command identifier and the command payload length anticipating contiguous
blocks of adr commands. If we find a discrepency in size, we should
abort.

Commit:
Updating unit tests

Updating unit test in response to the change in the link ADR related
APIs.

rajkan01 committed:
travis-ci fix for spell check

paul-szczepanek-arm committed:
Get the pal event handler from pal gattclient

paul-szczepanek-arm committed:
remove unused converter

paul-szczepanek-arm committed:
require pal event handler to be complete

paul-szczepanek-arm committed:
use grandfathered convention for class members names and initialise them

vmedcy committed:
PSOC6: enable export to CMake

The approach for the hex_files subset selection is identical
to makefile exporter: https://github.com/ARMmbed/mbed-os/pull/9466

Single hex file should be passed to srec_cat when hex_filename
is set in targets.json or mbed_app.json.

paul-szczepanek-arm committed:
move cpp to sources

Commit:
Storage mbed_lib.json files

Commit:
Add PSA mbed-lib.json

Commit:
[trusted-firmware-m]: Updated to 45e5276

Commit:
ESP8266: fixes UART HW flow control enabling sequence

Commit:
ESP8266: protects setting UART HW flow control with a mutex

Commit:
ESP8266: prevents doing reset twice in a row, unnecessary

Commit:
ESP8266: retry reset attempt once in failure

c1728p9 reviewed (changes requested):
This change needs to be made against the watchdog feature branch

bridadan committed:
Do not provide a default Pelion host address.

This moves the default pelion host address to be assigned by the cloud
sdk. It also cleans up the logic a bit to be shorter.

Commit:
TF-M patch: General modifications

- Remove un-needed files
- Disable printf and uart
- Modify include paths
- Guard macros from mbed_lib with ifndef

Commit:
TF-M patch: Fix tfm_ns_lock_init issue (TF-M issue #239)

- Link to bug tracking: https://developer.trustedfirmware.org/T239

Commit:
TF-M patch: Fix service handles not cleared issue (TF-M issue #230)

- Link to bug tracking: https://developer.trustedfirmware.org/T230

Commit:
TF-M patch: Fix fixed-size partitions stack size (TF-M issue #240)

- Link to bug tracking: https://developer.trustedfirmware.org/T240

Commit:
TF-M patch: Fix tfm_memory_check issue (TF-M issue #224)

- Link to bug tracking: https://developer.trustedfirmware.org/T224

Commit:
TF-M patch: Fix tfm_psa_call_venner wrong argument type (TF-M issue #241)

- Link to bug tracking: https://developer.trustedfirmware.org/T241

Commit:
Generate partition files

Commit:
TF-M patch: Fix wrong check in tfm_spm_check_client_version  (TF-M issue #236)

- Link to bug tracking: https://developer.trustedfirmware.org/T236

Commit:
Implement PSA protected storage & restructure PSA storage implementation

- Move all PSA storage code under psa/storage directory
- Create a global PSA error codes header, eliminating ITS specific ones
- Create a common header file for PSA storage type definitions,
  eliminating ITS specific ones
- Create a common implementation for PS & ITS
- Implement protected storage feature
- Change ITS test to be common to PS as well

Commit:
Change TESTS/netsocket/ for cellular testing

Commit:
Build SPM tests only for MBED_SPM targets

Commit:
Add TF-M autogen tool

Commit:
Add mbed_lib.json to TF-M SPE

Commit:
Update tfm_importer.json with SHAs of patches

Commit:
TF-M patch: Fix psa_set_rhandle() issue during CONNECT (TF-M issue #216)

- Link to bug tracking: https://developer.trustedfirmware.org/T216

paul-szczepanek-arm committed:
added spdx licence ident

Commit:
Cellular: clean sensitive information from trace

Information removed from the cellular subsystem trace log:
* SIM pin
* APN username and password

Added get_debug() to ATHandler.h to get the current trace state in order
to support suspend/resume of trace output.

ngraziano committed:
Vscode export: replace obsolete ${workspaceRoot}

Replace with ${workspaceFolder}

Commit:
revert: "Refactor sleep tracing driver identifier to be pointer to the driver filepath."

This reverts commit 9a0e8797b9b9fc32f6ce8f5fc508a43ff76b42f8.

Commit:
Update linker files and resolve build issue

Linker files for LPC11U68 and LPC1549 updated as Mbed OS memory model
1.	LPC1549
Has multiple memory banks, 4K bank used for stack exclusively and 16K for Heap
Its is mbed 2 only target hence 4K for stack.

2.	LPC11U68
Added Heap and Stack sections in linker files. Resolved build issues.

Commit:
Prefixed string to properly be parsed as bytes

Commit:
Removed decode since deeper tools always return strings

Commit:
Add MBEDTLS_ENTROPY_NV_SEED to NUCLEO_F411RE to fix the build

lrusinowicz committed:
FUTURE_SEQUANA_PSA: fixed deep sleep mode

Enabled tickless mode for Sequana PSA M0 core code to allow it to enter
deep sleep mode. This fixes issue #9094 where tests were failing due to
M0 core not entering deep sleep mode, blocking the whole chip.
Fixed incorrect resource management on M0 core, which crashed tickless
mode.

paul-szczepanek-arm committed:
fix nrf52 mtu

0xc0170 committed:
readme: license update

Fix licenses based on latest Fossology scan. Using -style licenses for specific type licenses
Add paths to the readme.

0xc0170 committed:
readme: storage license under one folder

lrusinowicz committed:
FUTURE_SEQUANA: Fixed Arduino signal mappings

Fixed Arduino signal mappings for production version of Sequana board.

KariHaapalehto committed:
enet_tasklet_disconnect() should generate event to tasklet.
enet_tasklet_disconnect() is now generating event to ethernet_tasklet,
which will then handle it.

NeilMacMullen committed:
Make a few trivial syntactic changes to better support cross-compilation

michalpasztamobica committed:
Greentea tests: set correct port to test against google.com

NeilMacMullen committed:
Keep style linter happy

c1728p9 committed:
Update pinmap design document

Update the pinmap design document to reflect the current design and to
match the design document template.

Commit:
btle: fix build warning about unused clockConfiguration

Fix the following build warning seen with both GCC and ARM_CC
[Warning] btle.cpp@115,0:  #177-D: variable "clockConfiguration"  was declared but never referenced

Commit:
target: UBLOX: Fix build warning in UBLOX_AT_CellularNetwork.cpp

Fix this build warning seen when building with ARMCC

Compile [ 13.7%]: UBLOX_AT_CellularNetwork.cpp
[Warning] UBLOX_AT_CellularNetwork.cpp@65,0:  #111-D: statement is unreachable

Commit:
platform: fix build warning in mbed_error

Fix the following build warning

Compile [ 83.4%]: mbed_error.c
[Warning] mbed_error.c@71,21: 'compute_crc32' defined but not used [-Wunused-function]

compute_crc32 usage is guraded with #define, but not the definition.
Use the same #define around the definition of compute_crc32()

Commit:
nanostack: icmpv6: fix build warning

Fix the following build warning seen when building with GCC

Compile [ 54.2%]: icmpv6.c
[Warning] icmpv6.c@1091,16: this statement may fall through [-Wimplicit-fallthrough=]

Commit:
features: nanostack: fix warning in thread_extension.h

Fix the following build warning seen with GCC

Compile [ 51.2%]: thread_bootstrap.c
[Warning] thread_extension.h@88,44: statement with no effect [-Wunused-value]

michalpasztamobica committed:
Netsocket documentation cleanup

Commit:
Fix dependencies of platform partition

0xc0170 committed:
license: rename to apache 2.0 txt file

0xc0170 committed:
license: add license file containing folders with non apache license

Commit:
Cellular: fix onboard modems powering failure

Don't call soft power on if device was already powered.
Some modems need to wait before sending anything to modem after soft powering.

Commit:
Refactor psa entropy inject test

Replace TEST_ASSERT with more specific assertions

Commit:
Clear ITS environment in psa entropy tests

mbed_psa_reboot_and_request_new_security_state() API replaced its_reset() which is now a secure API only

This change is necessary for a clean environment for the test

0xc0170 committed:
readme: license file reference fix

Commit:
Consolidating tools/tfm and tools/spm into tools/psa

Commit:
Reverting to original doxy configuration

This commit essentially brings back old code. We did an attempt to
resolve doxy issues with nfc durint docathlon, but it introduced more
issues and yak shaving rather than being a useful exercise.
Unfortunately the previous attempt got merged. This commit attempts to
restore the style maintained by original developers. From here they can
take over.

Commit:
Make targets plural in design document

Co-Authored-By: c1728p9 <butleja10511@hotmail.com>

Patater committed:
crypto: Remove PSA Crypto API from targets without entropy

An entropy source is required in order to use the PSA Crypto API. The
only devices Mbed OS knows are guaranteed by default to have an entropy
source are those devices with a TRNG. Don't enable the PSA Crypto API by
default for devices that Mbed OS can't know have an entropy source. This
avoids run-time errors when an entropy source is not present on these
targets.

Applications can add their own entropy source by place entropy into
their systems, implementing their own NV Seed read and write callbacks,
and then enabling the MBEDTLS_ENTROPY_NV_SEED configuration option to
notify the PSA Crypto implementation that an entropy source is present
and how to use it.

See https://os.mbed.com/docs/mbed-os/v5.11/porting/entropy-sources.html
for the background on why entropy is fundamental to system security and
how to inject entropy into systems that lack an on-board source of
entropy.

orenc17 committed:
Add release notes section to PR template

c1728p9 committed:
Update table of content links

Update the links and names in the table of contents.

Commit:
Only GCC_ARM toolchain is supported, removing legacy code

bridadan committed:
Add .inc as a header file type.

Currently, if you were to add a directory that only contained .inc
files, this file would not be passed as an include path to the compiler.
This ensures that .inc files are also considered header files.

Commit:
Modify license about file GPIO, SPI and system

Commit:
Add license in file cmsis_nvic.h

Commit:
Add mbedtls platform setup and teardown to modules

Add calls to `mbedtls_platform_setup()`
and `mbedtls_platform_teardown()` to all modules and tests using Mbed TLS.

paul-szczepanek-arm committed:
removed redundant qualifier tripping up GCC

0xc0170 committed:
readme: fix space and . in license

michalpasztamobica committed:
Greentea tests: user-configurable timeouts

Now it is enough to add:
    "macros": [
        "MBED_GREENTEA_TEST_XXXSOCKET_TIMEOUT_S=20"
    ],
to mbed_app.json, where XXX is on of {DNS, TLS, UDP, TCP}.
Also network-* tests are now configurable: network-interface, network-wifi, network-emac with a similar macro.

Commit:
Use a mutex in platform setup \ teardown functions

Use a singleton Mutex in platforms_alt functions, to be shared
with the trng function, to save RAM. Rename `platform_alt.c`
to `platform_alt.cpp` as the mutex is in a `singletonPtr`
template class.

0xc0170 committed:
cmsis conf: change license to apache 2.0

0xc0170 committed:
wifi test: add license info

0xc0170 committed:
test mbedmicro: fix licenses

0xc0170 committed:
check release: add license header

cmonr commented:
@rajkan01 Making sure, is the intention of this PR to test the new branch before pushing it, or to merge changes into the watchdog feature branch?

@c1728p9 You might be able to help clarify as well.

Commit:
Add flash sum check

Check that flash size sum up to all sectors combined

Commit:
tools/utils: ensure default build profiles are processed first

The command-line argument '--profile' looks for build
profiles provided in mbed-os/tools/profiles/. If a
directory name exists in the root folder with the
same name as one of the profile names provided by
default [e.g debug/develop/release], that directory
is processed instead resulting in incorrect behavior.
Fix this behavior by processing the default profiles first.

Commit:
Add mbed_app.json for TF-M bare-metal build

Commit:
Define delivery directory for FUTURE_SEQUANA_PSA

bridadan committed:
Remove pyocd requirement and update icetea.

Icetea depends on mbed-flasher. mbed-flasher recently released a version
that drops the dependency on pyocd. Icetea released a version that uses
this latest version of mbed-flasher. So now tht pyocd is no longer in
our dependency tree, we can drop it here.

Commit:
Update .travis.yml psa_autogen job with TF-M code generation tool

Commit:
Cellular: Fix sigio to be released in ATHandler destructor

michalpasztamobica committed:
Greentea: Remove TLSSocket_simultaneous test

The reason is that some platforms lack memory to run it. To be brought
back once TLS library memory consumption is optimized.

Commit:
Re-generate TF-M partition files

vmedcy committed:
PSOC6: update PDL to the latest version

vmedcy committed:
PSOC6: minor updates to Cypress HAL

vmedcy committed:
PSOC6: do not reconfigure BT GPIO pins in CYW43XXX HCIDriver

0xc0170 committed:
tools: add license headers

0xc0170 committed:
copyright 2019 fix

michalpasztamobica committed:
ESP8266: connect() returns OK in non-blocking calls

Commit:
tools: export: MCUXpresso: fix case inconsistencies in .templ file names

The templ file names are both all upper case and lower case letters.
The Target Names map is usually all upper case. The match could fail
if the templ file, as we have case-sensitive comparison. Handle such
cases by perorming a case-insensitve check.

mbed export of a project to MCUXpresso could potentially always fail
irrespective of what is passed in -m option since the target names
map entry and the filename may not match. This commit fixes this issue.

Example of the issue that this commit fixes:

$ mbed export -i mcuxpresso -m lpc11u68 -v
<snip>
project.py: error: LPC11U68 not supported by mcuxpresso
<snip>

vmedcy committed:
PSOC6: update BSP generated sources with latest configurator

Use ModusToolbox Device Configurator 1.1.0.284 to generate the
BSP low-level initialization code. Compatible version of Device
Configurator to be released with ModusToolbox 1.1.

Notable changes:
* rename cycfg_connectivity -> cycfg_routing
* switch LF_CLK clock source from ILO to WCO on
  CY8CPROTO-062-4343W and CYW943012P6EVB-01

Commit:
Fix typo in LICENSE

Fix typo as requested in comments.

Commit:
Add articles to README.md

Include articles for clarity.

Commit:
Fix incorrect return value in psa_security_lifecycle_state (emul)

psa_security_lifecycle_state should return uint32_t of the security
lifecycle state.
bug: psa_platfrom_lifecycle_get_impl return value
is psa_status_t.

Commit:
Cellular: retire CellularBase class

It was decided to retire CellularBase class which served as a pure
virtual interface class from which Cellular network stack
implementations would get inherited. However, the current view is that
we may be the only user of it so we could retire CellularBase.

Commit:
TF-M patch: More fixes to tfm_memory_check issue (TF-M issue #224)

- Link to bug tracking: https://developer.trustedfirmware.org/T224

Commit:
Cellular: ATHandler debugging relieved

Commit:
Cellular: Handling of TCP endpoint closed

Commit:
Cellular: Fix Quectel M26 drivers

Commit:
Cellular: Fix Quectel BG96 driver RTS/CTS setup

Commit:
Cellular: Fix CellularDevice mutex unlock

Commit:
Cellular: Fix Gemalto/Cinterion socket open defer

Commit:
support in more types then just RSA_KEYPAIR

if the type is not supported the psa_crypto will return the error code
for unsupported

0xc0170 committed:
tools: fix hashbang

Patater committed:
crypto: test: Use new PSA error codes

The PSA Framework and PSA Crypto error code names have been unified and
renamed in some cases. Use the new unified names.

Patater committed:
crypto: Update Mbed Crypto to 1.0.0d3

Commit:
Fix SPM returning the wrong value

Commit:
Add USER_BUTTON and SPI in file PinNames.h

desmond-blue committed:
BLE:Cordio: ATTS setting for write cback should use bitwise or

Commit:
Cellular: Fix ATHandler set/restore_at_timeout with mutex lock

Commit:
Cellular: Added explanatory comment for AT cellular stack's socket_open routine

Commit:
Cellular: add SPDX identifier

0xc0170 committed:
memap: remove not needed comment prior future import

Commit:
Update Nanostack_HAL licenses and copyright year

-Add Apache License to files that were missing a licence
-Update copyright year

Commit:
Update copyright years

Commit:
Cellular: Remove compiling warning

0xc0170 committed:
tools: move string description after import modules

Commit:
allow calling hash or mac on large buffers without large memory allocation

use a fixed size buffer for large inputs to process the input
in multiple smaller pieces.

Commit:
Add BSD3-clause license file to TF-M folder

Commit:
TFM: Add defines to mbed_lib.json

tymoteuszblochmobica committed:
Multihoming change parameter orded in gethostbyname

Commit:
Rebuild new Wiced binaries

paul-szczepanek-arm committed:
fixed initialisation order

vmedcy committed:
tools/export: move hex_files selection to base Exporter class

CMake and makefile exporters share a common logic for hex file
selection. Factor it as a common property in the base class
to avoid code duplication.

Commit:
Add config header to assembly compilation

ARMC6 was not including mbed_config.h when compiling assembly files

TacoGrandeTX committed:
Remove ITM from NRF52_DK and DELTA_DFBM_NQ620 targets

The NRF52_DK and DELTA_DFBM_NQ620 have the SWO pin (p0_18) mapped to
LED2. This means that on startup LED2 turns on after the ITM is
initialized which is confusing. Since most users want LED2 usage
instead of SWO we remove the ITM for these targets.

The nRF52 readme is updated to instruct users how to get SWO support
if they need it.

Commit:
Cellular: fix for binary compability requirement for CY8CKIT_062_4343W

0xc0170 committed:
license: use .md extension

Using Github feature to show markdown documents.

Commit:
TFM Importer: Update SHAa of TFM patches

michalpasztamobica committed:
Multiple memory handling fixes

Correct memory clean-ups in multiple tests and a stub.

michalpasztamobica committed:
Multiple memory handling fixes

Based on valgrind reports running on unit tests following changes were introduced:
* TLSSocketWrapper frees allocated cert buffer in case of errors from mbedtls,
* nsapi_addr has a mem_init() function, initializing all of its memory during construction.

vmedcy committed:
PSOC6: fix port_write API

Fix port_write API to correctly shift the passed value.
This allows the reference application provided in PortOut docs
to work corectly with arbitrary LED_MASK.
https://os.mbed.com/docs/mbed-os/v5.11/apis/portout.html

The fix applies to both PSOC6 and PSOC6_FUTURE HAL implementations.

guialonsoalb committed:
Added new target RHOMBIO_L476DMW1K

ConradBraam committed:
adv builder helper now returns an object

Commit:
Adding STM32_L475 as a new MTB target

Commit:
Adding SPDX license identifier

Commit:
Adding SPDX license identifer for all new files

michalpasztamobica committed:
Add missing wifi parameters to test configs

Commit:
Removing redundant code comments

guialonsoalb committed:
added licence identifier texts

Patater committed:
crypto: Update to Mbed Crypto 1.0.0d4

vmedcy committed:
PSOC6: correctly align hex files with split text sections

When the original PSOC6 CM4 hex file contains unalinged text sections
that span through multiple intelhex segments, aligned segments (filled
with zeroes) overlap with the original data segments, resulting in
error thrown by ihex.merge(alignments, overlap='error').
Such hex file can be produced when the ELF is built with ARM MDK Compiler
with --split_sections option:
http://www.keil.com/support/man/docs/armcc/armcc_chr1359124944914.htm

Change the merge strategy to overlap='ignore', so that the overlapping
zero-filled segments are skipped.

bridadan committed:
Ensure header CRC is written as unsigned int

kegilbert committed:
Change EventFlag timeout paramter

Matches rest of RTOS class timeout parameters by using the unit name. Remove ambigious statement in reference to 0 ms being no-timeout as a timeout of 0 causes the function to not block and return immediately (osWaitForever is used as no timeout as it will wait forever)

Commit:
Add ARM_LIB_STACK to resolve CI Error

Commit:
Update crypto IPC handling of negative errors

Commit:
Add flash IAP to sequana_psa m4

michalpasztamobica committed:
Document network parameters in netsocket/mbed_lib.json

Commit:
Changing SPDX license identifier to BSD-3-Clause

guialonsoalb committed:
Change in license identifier, removed commented dead code

Change in license identifier
Removed commented dead code
Removed some tab space

sarahmarshy committed:
Support mbed_start_application for Cortex-M0+

Commit:
Add image signing scripts from TF-M bl2 library

michalpasztamobica committed:
Refactor tcpsocket tests to use sigio and not to wait

Commit:
New Wiced binaries

jeromecoutant committed:
STM32 LPTICKER (LPTIM): correct init execution after bootloader

sarahmarshy committed:
Reuse nvic powerdown code block for M0+

Commit:
Fix psa_system_reset on PSA IPC platform

SenRamakri committed:
Update mbed_stats.c to use __ARMCC_VERSION flag to be compatible with AC6

orenc17 committed:
Update secure binaries

Commit:
TF-M patch: Free connection handle in psa_reply (TF-M issue #263)

- Link to bug tracking: https://developer.trustedfirmware.org/T263

paul-szczepanek-arm committed:
put cordio cfg in mbed_lib.json

paul-szczepanek-arm committed:
mention config in readme

sarahmarshy committed:
Change MOV to MOVS for M0+

sarahmarshy committed:
Change to MOVS for all targets

paul-szczepanek-arm committed:
prepared writes options

cydriftcloud committed:
PSOC6: send HCI reset after patch RAM download

After the patch RAM download is completed, a HCI reset should be sent in order to initialize the registers. Some of the initialization won't be called if the HCI reset is not sent after firmware download.

Commit:
return the new values for key type & size even if function fails

Commit:
Remove unnecessary prints in FileSystemStore class

Commit:
a-style fixes for code

Commit:
fix comment issues

Commit:
a-style fix for unrelated code

Commit:
Add api about I2C and SPI ....

Commit:
reopening #9754 after the changes where overrun
changing the SPM error codes from invalid argument to bad state aligned to PSA spec

paul-szczepanek-arm committed:
use existing macro names

paul-szczepanek-arm committed:
revert macros

teetak01 committed:
Update mbed-cloud-sdk to latest version 2.0.4

Commit:
Remove Arm Beetle target

Deprecation in preparation for Mbed OS 5.12 and move to IAR8 and ARMC6

Commit:
Add missing copyright header to STM32 driver

Commit:
Replace Copyright ARM with a proper Apache 2 license header

Commit:
TFM Importer: Update SHAa of TFM patches

michalpasztamobica committed:
Change all wifi and WiFi occurences to Wi-Fi

paul-szczepanek-arm committed:
stray comma

paul-szczepanek-arm committed:
fixed names

paul-szczepanek-arm committed:
typos

d-kato committed:
Add missing license description to RZ/A1 driver

rajkan01 committed:
fix: Fixing rebased created bugs

rajkan01 committed:
update:Remove DEVICE_WATCHDOG and added RESET_RESON in stm32 family
